aws/logs_monitoring/parsing.py

```diff
@@ -863,3 +853,42 @@ def get_state_machine_arn(message):
         arn_tokens[5] = "stateMachine"
         return ":".join(arn_tokens[:7])
     return ""
+
+
+# Lambda logs can be from either default or customized log group
+def process_lambda_logs(logs, aws_attributes, context, metadata):
+    lower_cased_lambda_function_name = get_lower_cased_lambda_function_name(logs)
+    if lower_cased_lambda_function_name is None:
+        return
+    # Split the arn of the forwarder to extract the prefix
+    arn_parts = context.invoked_function_arn.split("function:")
+    if len(arn_parts) > 0:
+        arn_prefix = arn_parts[0]
+        # Rebuild the arn with the lowercased function name
+        lower_cased_lambda__arn = (
+            arn_prefix + "function:" + lower_cased_lambda_function_name
+        )
+        # Add the lowe_rcased arn as a log attribute
+        arn_attributes = {"lambda": {"arn": lower_cased_lambda__arn}}
+        aws_attributes = merge_dicts(aws_attributes, arn_attributes)
+        env_tag_exists = (
+            metadata[DD_CUSTOM_TAGS].startswith("env:")
+            or ",env:" in metadata[DD_CUSTOM_TAGS]
+        )
+        # If there is no env specified, default to env:none
+        if not env_tag_exists:
+            metadata[DD_CUSTOM_TAGS] += ",env:none"
+
+
+# The lambda function name can be inferred from either a customized logstream name, or a loggroup name
+def get_lower_cased_lambda_function_name(logs):
+    logstream_name = logs["logStream"]
+    # function name parsed from logstream is preferred for handling some edge cases
+    function_name = get_lambda_function_name_from_logstream_name(logstream_name)
```

Comment:
Rather than run this regex match a second time for every lambda log event, does it make sense to set the `function_name` in `awslogs_handler` based on whether we've discovered it to be a custom log group or not?

Reply:
I was thinking of avoiding pattern matching twice, and it turns out that in order to be compatible with the existing logics for default log group, the codes can be more verbose and hard to maintain. I tried to refactor the lambda log processing logic out into a separate function at the cost of it.